### ISSUE TITLE
RPC: add ability to subscribe to multiple accounts notifications - `accountsSubscribe`

### ISSUE THREAD

**Issue by thaaddeus** [solana-labs/solana#12237]
#### Problem
Currently it's not possible to subscribe to multiple accounts updates via single subscription request and receive `accountsNotification` via WS. 

Only way is to use `accountSubscribe` and subscribe via separate requests, but downside of that is that the order of received notifications for multiple accounts is not guaranteed in any way.

#### Proposed Solution
Add `accountsSubscribe` that would accept array of accounts PubKeys and send `accountsNotification` when any of the 'tracked' accounts changes for given slot.

**Comment by thaaddeus** [comment]
@jstarry what do you think? It would allow clients that use multiple accounts data (like serum dex) to have consistent/synced view on that data when notified.

**Comment by jstarry** [comment]
Hey @thaaddeus yeah this makes a lot of sense to me

**Comment by thaaddeus** [comment]
awesome, thanks! I don't know rust that much yet, but anyway will try to contribute somehow.

**Comment by jstarry** [comment]
That'd be great! I think it's probably cleanest to add server support for this

I should note that it is possible to implement this now in the client / web3 sdk without any server changes but at a slight delay. This is possible because account notifications on the pubsub api already return the context slot for which the account last changed. The client can track the state of accounts per slot but if one of the subscribed accounts *didn't* change in a slot, it's tricky to be certain about that.

**Comment by thaaddeus** [comment]
@jstarry actually I did implement it client side with delay as you described but not sure if it's correct due to https://github.com/solana-labs/solana/issues/9910 - in essence when using `singleGossip` commitment can I assume that `context.slot` for given notification is good "key" to group notifications for? or that should be `last_modified_slot` as described in linked issue?

**Comment by jstarry** [comment]
Sorry, my past self is correct here 😅  `last_modified_slot` is needed to do client side syncing

**Comment by thaaddeus** [comment]
awesome, thanks for confirming this as it was driving me crazy why it sometimes break (synching based on `slot`).

**Comment by jstarry** [comment]
Hey @thaaddeus, I'm looking into this now, sorry for the wait!

**Comment by thaaddeus** [comment]
Great, really appreciate it, thank you!

**Comment by jstarry** [comment]
@thaaddeus after some more digging I think I have a more accurate view of what's going on here and how you can reason about the current API.

1. Single Gossip Reliability

Before v1.3.14, the `singleGossip` commitment level was unreliable for account subscriptions. This meant that it was possible to sometimes receive delayed notifications or notifications for account state mid-block. I'm pretty sure this was the main cause of your syncing issues. This has been fixed on mainnet since early October.

2. Account notification context slot

> `last_modified_slot` is needed to do client side syncing

This was wrong. You don't need the last modified slot for syncing. The context slot is sufficient because at that slot, we look at a snapshot of all accounts and send updates for each account that has changed. If an account doesn't notify for that slot, you can be assured that it hasn't been updated since the last notification. So, after processing the notifications for a given context slot, you can be confident you have an accurate synced view of the accounts you care about at the context slot.

3. Multiple account notifications

Since client syncing is a pain and requires a slight delay to make sure you got all the notifications for a slot. I'm working on this now!

**Comment by thaaddeus** [comment]
@jstarry ah I see, indeed it was happening also with `singleGossip` (with `recent` it was more frequent), makes sense now, thanks!

**Comment by jstarry** [comment]
@thaaddeus can you describe what behavior you were seeing with `recent`?

**Comment by thaaddeus** [comment]
I think it was the same as with `singleGossip` just more frequent, but I'm not 100% sure anymore, I'll try to reproduce it today both with `recent` and `singleGossip` (here it should not happen anymore as you've described).

**Comment by jstarry** [comment]
Great, thank you!

**Comment by thaaddeus** [comment]
I'm running it now and can confirm that so far with `singleGossip` and client side syncing it works perfectly fine, thanks!
When I switch to `recent` I can see the issue, but I think it's not related to that, but simply an issue due to block rollbacks - I guess `recent` is not a good choice for my use case.

**Comment by jstarry** [comment]
Awesome, that's expected, thanks for checking! Just to be clear, singleGossip will also get rolled back occasionally so you should handle that case

**Comment by thaaddeus** [comment]
@jstarry I got it working pretty much fine with client side syncing - client waits up to 10 seconds for remaining accounts notifications, but even with 10s delay it's not enough which is bit unusual from my perspective as notifications for given slot should be published around the same time , > 10s is quite a lot - I'm running it all as a local node to rule out networking issues and with `confirmed` commitment level. Anyways having support for multiple accounts notification via WS would be great add-on.

**Comment by jstarry** [comment]
> @jstarry I got it working pretty much fine with client side syncing - client waits up to 10 seconds for remaining accounts notifications, but even with 10s delay it's not enough which is bit unusual from my perspective as notifications for given slot should be published around the same time , > 10s is quite a lot - I'm running it all as a local node to rule out networking issues and with `confirmed` commitment level. Anyways having support for multiple accounts notification via WS would be great add-on.

Interesting, yeah 10s is really long. Are these accounts really big?

**Comment by thaaddeus** [comment]
actually yes, serum `eventQueue` accounts are ~1MB.

**Comment by jstarry** [comment]
In that case, I think that reducing websocket bandwidth would have bigger impact.

**Comment by thaaddeus** [comment]
thing is currently in order to get real-time trade data for serum you need to check `eventQueue` account, it's all local network on single dedicated host, so I not sure what else can be done except querying `eventQueue` periodically, but then it's not real-time anymore.

**Comment by jstarry** [comment]
Yeah, I think we could solve it by sending over account patches instead of full account data each time

**Comment by thaaddeus** [comment]
That would be great indeed!

**Comment by israelidanny** [comment]
I opened an issue for sending account patches:
https://github.com/solana-labs/solana/issues/17496

**Comment by dave-fl** [comment]
Delay is mentioned several times here - what is a reasonable delay? 10s seems quite high.
